### ISSUE TITLE
P2P: Node with onlynet=ipv6 connects outgoing to Onion

### ISSUE THREAD

**Issue by wodry** [bitcoin/bitcoin#13378]
Running bitcoind 0.16.1rc1 64-Bit binary release on Linux

### What I did
I have now setup IPv6 connection for my node (but only outgoing, since my providers router does not allow incoming IPv6 traffic...)
I have running also Tor on localhost with bitcoin (not really) "hidden" service, in parallel to IPv4.

So I was curious and set 
onlynet=ipv6
in bitcoin.conf (before, "onlynet" was not set)

### Expected behavior
Only IPv6 connections, both incoming and outgoing, no Tor traffic, no traffic to local Tor hidden service.

### Actual behavior
Node connects to IPv6, IPv4 and Tor, see screenshot of GUI debug window connection tab.

Also, there are error messages in the debug log that I had not before:
("Keine Berechtigung" means "No permission" in English)

2018-06-03 12:41:46 Loaded 71277 addresses from peers.dat  324ms
2018-06-03 12:41:46 init message: Lade Sperrliste...
2018-06-03 12:41:46 init message: Starte Netzwerk-Threads...
2018-06-03 12:41:46 init message: Laden abgeschlossen
2018-06-03 12:41:46 opencon thread start
2018-06-03 12:41:46 net thread start
2018-06-03 12:41:46 dnsseed thread start
2018-06-03 12:41:46 addcon thread start
2018-06-03 12:41:46 msghand thread start
2018-06-03 12:41:46 GUI: Platform customization: "other"
2018-06-03 12:41:46 GUI: PaymentServer::LoadRootCAs: Loaded  159  root certificates
2018-06-03 12:41:55 Imported mempool transactions from disk: 2019 succeeded, 0 failed, 0 expired, 0 already there
2018-06-03 12:41:57 Loading addresses from DNS seeds (could take a while)
2018-06-03 12:41:57 219 addresses found from DNS seeds
2018-06-03 12:41:57 dnsseed thread exit
2018-06-03 12:42:07 New outbound peer connected: version: 70015, blocks=525773, peer=0
**2018-06-03 12:44:06 GUI: QAccessibleTable::child: Invalid index at:  3 0
2018-06-03 12:44:06 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  29
2018-06-03 12:44:06 GUI: QAccessibleTable::child: Invalid index at:  3 0
2018-06-03 12:44:06 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  29
2018-06-03 12:44:06 GUI: QAccessibleTable::child: Invalid index at:  3 0
2018-06-03 12:44:06 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  29
2018-06-03 12:44:11 GUI: QAccessibleTable::child: Invalid index at:  2 0
2018-06-03 12:44:11 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  22
2018-06-03 12:44:11 GUI: QAccessibleTable::child: Invalid index at:  2 0
2018-06-03 12:44:11 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  22
2018-06-03 12:44:11 GUI: QAccessibleTable::child: Invalid index at:  2 0
2018-06-03 12:44:11 GUI: Cannot creat accessible child interface for object:  QTableView(0x55960b4b1040, name = "peerWidget")  index:  22
2018-06-03 12:44:51 connect() to [2601:98a:4203:6e9d:68b9:19cf:e683:6932]:8333 failed after select(): Keine Berechtigung (13)
2018-06-03 12:45:33 connect() to [2a00:c1a0:4888:a500:cc5b:3e64:8aba:d6b4]:8333 failed after select(): Keine Berechtigung (13)
2018-06-03 12:45:38 connect() to [2001:981:9cae:1:7c4b:8f56:ab7c:78f]:8333 failed after select(): Keine Berechtigung (13)**

![bildschirmfoto](https://user-images.githubusercontent.com/8447873/40886866-c73f5f9a-673f-11e8-8100-50cf24168067.png)

**Comment by wodry** [comment]
When setting
onlynet=ipv4
there are outgoing connections to .onion addresses, too.

The bitcoind man page says:

> -onlynet=<net>
>       Only connect to nodes in network <net> (ipv4, ipv6 or onion)

Additionally, I get incoming connections via Tor Hidden Service with both onlynet=ipv6 and onlynet=ipv4.

If this should be not a bug, but a feature, the man page should explain this behaviour more detailed and precisely.

I am wondering, that there is no feedback yet about this behavior. I think it is easy reproducable.

**Comment by sipa** [comment]
Onlynet only affects outgoing connections, not incoming.

**Comment by wodry** [comment]
Thanks for feedback. Following this, I would suggest change the man page text to:

```
onlynet=
Actively connect to remote nodes exclusively in network <ipv4|ipv6|onion>.
Passive, incoming connections are not affected by this parameter.

```
But this statement for active connections to remote nodes does not seem to match the reality, currently. Or am I missing something?

**Comment by sipa** [comment]
> Thanks for feedback. Following this, I would suggest to change the man page text to:

Sounds good to me. Feel free to open a PR.

> But this statement for active connections to remote nodes does not seem to match the reality, currently. Or am I missing something?

If that's the case, there is a bug, and we should keep this issue open to investigate. I was just commenting because the in/out distinction was a possible explanation for what you saw.

**Comment by kallewoof** [comment]
@sipa 
> Onlynet only affects outgoing connections, not incoming.

Huh. So a `onlynet=tor` user may end up getting non-tor incoming connections? That sounds potentially undesirable.

**Comment by wodry** [comment]
Is https://github.com/bitcoin/bitcoin/pull/13418 ready for merge?

**Comment by wodry** [comment]
I got the node now running for 48h with `onlynet=ipv6`, and from looking at the connections tab from time to time, I only once saw a single outgoing connection via IPv6, so pretty seldomly. All of the other outgoing connections are to Onion addresses. I never saw an outgoing connection via IPv4.

**Comment by wodry** [comment]
`netstat -np | grep tor` shows only IPv4 connections, allthough Tor is configured with `ClientUseIPv6 1`,  so it looks like this is not a "Onion via IPv6" phaenomen.

**Comment by wodry** [comment]
Issue still exists with 0.21.0rc2, see screenshot of connection table running with `onlynet=ipv6`.

![bildschirmfoto](https://user-images.githubusercontent.com/8447873/101277624-115fd200-37b6-11eb-83c6-a6e85318d326.png)

**Comment by wodry** [comment]
There are no outgoing onion connections when additionally to `onlynet=ipv6` setting `listenonion=0` (default: 1).

**Comment by wodry** [comment]
Check if this is solved by #22651.
